### PR TITLE
Known as

### LINKED ISSUE #84: Failing to extract definitions form sentences with "known as" construction
It looks like we fail to get definitions from sentences with AKA clauses ("known as").

Example: "Arlotto Mainardi, known variously as Pievano Arlotto or Piovano Arlotto, was a priest known for jests and "pleasantries." List of subjects: `([u'Arlotto Mainardi', u'Pievano Arlotto', u'Piovano Arlotto', u'Piovano Arlotto'])`

Other examples:

```
MainThread - DEBUG -   .. whoami exception for: "Norma Sue Edwards Scott, known as Sue Scott, is a Republican member of the Arkansas House of Representatives from Rogers in Benton County in northwestern Arkansas." ([u'Sue Scott', u'Norma Sue Edwards Scott', u'Sue Scott (politician)'])

MainThread - DEBUG -   .. whoami exception for: "Indalecio León Zarasqueta Uriarte, known as Txikito de Eibar or Aizpiri txiki, was a Basque pelota player who participated in "pala", Hand-pelota, "remonte", long bat and short bat categories." ([u'Txikito de Eibar', u'Aizpiri txiki', u'Indalecio Le\xf3n Zarasqueta Uriarte', u'Indalecio Sarasqueta'])

MainThread - DEBUG -   .. whoami exception for: "Club Atlético Kimberley known as Kimberley de Mar del Plata is a sports club from Mar del Plata, Buenos Aires Province, Argentina." ([u'Club Atl\xe9tico Kimberley', u'Kimberley de Mar del Plata', u'Kimberley de Mar del Plata'])

MainThread - DEBUG -   .. whoami exception for: "Lawrence Fewell Roberts II, known as Robin Roberts and in his music career as "Rockin' Robin" Roberts, was an American singer best known for his performances in the early 1960s with The Wailers, a rock and roll band based in Tacoma, Washington." ([u'Robin Roberts', u'Lawrence Fewell Roberts II', u"Rockin' Robin Roberts"])

MainThread - DEBUG -   .. whoami exception for: "Rebuild of Evangelion, known in Japan as , is a Japanese animated film series and a reboot of the original "Neon Genesis Evangelion" anime TV series, produced by Studio Khara." ([u'Rebuild of Evangelion'])
```

### PR DESCRIPTION
**Pull request by shu98**
Resolves #84

### CODE REVIEW

**Comment by shu98** [comment]
@pondpremt  I wasn't completely sure how to approach this one. What I ended up doing is just adding another function (handle_known_as) that modifies the sentence such that the "known as" form is dealt away with. This leaves the sentence in the usual _is-a_ form so that spacy can parse it correctly. For example 

_Alyssa is known as a spy. --> Alyssa is a spy._ 

_Norma Sue Edwards Scott, known as a congresswoman, is a Republican politician. --> Norma Sue Edwards Scott, a congresswoman, is a Republican politician._ 

_Argus Filch, known in the wizarding world as a Squib, is the caretaker of Hogwarts. --> Argus Filch, a Squib, s the caretaker of Hogwarts._ 

The only thing that is kind of awkward is something like this: 

_Norma Sue Edwards Scott, known as Sue Scott, is a Republican politician. --> Norma Sue Edwards Scott, Sue Scott, is a Republican politician._ 

Unfortunately, this function is kind of long and ugly, and probably not super elegant. I added it to a new file called util.py since I expect that as I am working through more issues, I may add more similar functions that handle small bugs. In any case, it is available for you to review. 

Please let me know if this is an appropriate way to go, or if you have suggestions for better ways to handle these bugs. Since we are using spacy to do the heavy-lifting parsing and I am not unfortunately not a linguistics expert (working on it!), my current solution is to just edit sentences into a form that can be parsed easily. Thank you!

**Comment by shu98** [comment]
Another issue that this brings up is [appositives](https://github.com/infolab-csail/whoami/issues/61). Since in this sentence, "Norma Sue Edwards Scott, a congresswoman, is a Republican politician", whoami doesn't return "congresswoman" as a definition.

**Comment by pondpremt** [comment]
I'm trying to understand the problems here. I see 3 separate issues from the comment above:

- In the construction "Subject is known as NP" we cannot pick out the NP as a definition because "known as" blocks the NP from being the direct attr-child of the main verb. The proposed approach works. For example, "Alyssa is known as a spy."
- In the construction "Subject, known as NP1, is NP2" where NP1 isn't a proper name, we want to extract NP1 as well. The current approach turns NP1 into an appositive, but whoami cannot pick out that appositive as a definition. For example, "Norma Sue Edwards Scott, known as a congresswoman, is a Republican politician."
- In the construction "Subject, known as NP1, is NP2" where NP1 is a proper name, we want to ignore NP1. The current approach works, but if we make whoami able to pick out the appositive it may incorrectly returns NP1 as a definition. For example, "Norma Sue Edwards Scott, known as Sue Scott, is a Republican politician."

** still editing **

**Comment by shu98** [comment]
Resolves #84

**Comment by pondpremt** [comment]
Assigning this back to @shu98 now, so she can do some more investigation of whether this is the right approach. 

Once you're done, please update the PR description, which is currently blank. A summary of the comments we have been making here + the final approach would do. Also, comments like "Resolves #84" should also be in the PR description rather than in a comment.

**Comment by shu98** [comment]
@pondpremt I have been doing more analysis of this issue, and believe that it may be safer to try resolving this issue through looking at the dependency tree rather than preprocessing the sentence. When I analyze the parses of various sentences, I get mixed results.

For the "<Subject>, known as <synonym or other definition>" pattern, the two main cases are:
- spaCy does not parse the sentence correctly when we substitute in the subjects (i.e. subject_number_00, etc.)
- the current version of spaCy doesn't parse the sentence correctly but the latest version does  

To me, this implies that we can fix these issues by upgrading spaCy and allowing whoami to try parsing both the substituted and non-substituted sentence. What are your thoughts?

For sentences in the "<Subject> is/was/etc. known as <definition>" form, neither version of spaCy parses the sentence correctly because they mark "known" as the root. I am working on figuring out how to deal with this.

**Comment by shu98** [comment]
@pondpremt Are there any test cases in particular that you would like me to try/add? I went through a number of example sentences and this approach seems to work. There is one scenario which I found this approach to sometimes fail. Consider the following sentence: 

> The asteroid known as Hermes is an eccentric, sub-kilometer sized asteroid and binary system.

The modified version (without the known-as clause) is

> The asteroid Hermes is an eccentric, sub-kilometer sized asteroid and binary system.

spaCy parses both the modified and unmodified sentences incorrectly (labels "asteroid" as the root), which leads to me to believe that if we try a different approach where we traverse the dependency tree, this example will still fail. I am not sure if whoami should be able to deal with it, but I'm inclined to think that (at least for now) it is a no.

File: whoami/preprocess.py [hunk 1]
```diff
@@ -107,6 +108,26 @@ def remove_parens(sentence):
...
+    source for the KGB. --> Melita Norwood was an intelligence source
+    for the KGB.
+
+    Arguments:
+    sentence (unicode) -- original sentence string
+
+    Return:
+    Sentence with 'known as' clauses removed
+    """
+    return re.sub(" known(.*?) as ", ' ', sentence)
```
Comment (shu98) [inline whoami/preprocess.py]: @pondpremt For now, I will leave it like this. After extensive testing, this approach seems to work with most Wikipedia sentences that contain 'known as'. I am still trying to figure out how to deal with more complex patterns, but I think we can merge this for now. Thoughts?

Comment (pondpremt) [reply whoami/preprocess.py]: Sounds good

**Comment by hanryhu** [comment]
It doesn't seem like this PR is needed anymore.  Issue https://github.com/infolab-csail/whoami/issues/84 appears to have been fixed by the recent spaCy update.

**Pull request closed without merging**